File: python/clock/clock.py
#!/usr/bin/env python3.4
""" Provides a Clock object to be used to track the time"""
# clock.py by Jack Hayhurst
#


class Clock:
    """ Clock is used to track the time, and allows minutes to be added or removed"""

    def __init__(self, hour=0, minute=0):
        self._hour = hour
        self._minute = minute
        self.clean_time()

    def clean_time(self):
        """cleanTime brings the hours and minutes on the clock into proper bounds"""
        while self._minute >= 60:
            self._minute -= 60
            self._hour += 1
        while self._minute < 0:
            self._minute += 60
            self._hour -= 1
        while self._hour < 0:
            self._hour += 24
        while self._hour >= 24:
            self._hour -= 24

    def __str__(self):
        return "{:02d}".format(self._hour) + ":" + '{:02d}'.format(self._minute)

    def add(self, minute):
        """add allows you to add or subtract minutes to/from the clock"""
        self._minute += minute
        self.clean_time()
        return str(self)

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return False

File: go/clock/clock.go
// Package clock provides a type clock that manually tracks a given time.
package clock

import "fmt"

const testVersion = 4

// Clock provides a time tracking clock.
// This clock does not actually track time and self update over time.
// Instead, minutes can be manually added/removed.
type Clock struct {
	hour   int
	minute int
}

// New creates a new clock and normalizes it
func New(hour, minute int) Clock {
	newClock := Clock{hour: hour, minute: minute}
	newClock.update()
	return newClock
}

// String prints out the current numbers on the clock
func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", c.hour, c.minute)
}

// Add adds the given minutes to the clock, then normalizes it
func (c Clock) Add(minutes int) Clock {
	c.minute += minutes
	c.update()
	return c
}

func (c Clock) Subtract(minutes int) Clock {
	return c.Add(-1 * minutes)
}

// internal function that brings the clock back into normal bounds
func (c *Clock) update() {
	for c.minute < 0 {
		c.minute += 60
		c.hour--
	}
	for c.minute >= 60 {
		c.minute -= 60
		c.hour++
	}
	for c.hour < 0 {
		c.hour += 24
	}
	for c.hour >= 24 {
		c.hour -= 24
	}
	return
}
